test/integ/aio/test_connection_async.py

```diff
@@ -18,8 +73,1443 @@ async def test_basic(db_parameters):
+
+
+@pytest.mark.skipolddriver
+async def test_with_tokens_expired(conn_cnx, db_parameters):
+    """Creates a connection using session and master token."""
+    try:
+        async with conn_cnx(
+            timezone="UTC",
+        ) as initial_cnx:
+            assert initial_cnx, "invalid initial cnx"
+            master_token = initial_cnx._rest._master_token
+            session_token = initial_cnx._rest._token
+
+        with pytest.raises(ProgrammingError):
+            token_cnx = snowflake.connector.aio.SnowflakeConnection(
+                account=db_parameters["account"],
+                host=db_parameters["host"],
+                port=db_parameters["port"],
+                protocol=db_parameters["protocol"],
+                session_token=session_token,
+                master_token=master_token,
+            )
+            await token_cnx.connect()
+            await token_cnx.close()
+    except Exception:
+        # This is my way of guaranteeing that we'll not expose the
+        # sensitive information that this test needs to handle.
+        # db_parameter contains passwords.
+        pytest.fail("something failed", pytrace=False)
+
+
+async def test_keep_alive_true(db_parameters):
+    """Creates a connection with client_session_keep_alive parameter."""
+    config = {
+        "user": db_parameters["user"],
+        "password": db_parameters["password"],
+        "host": db_parameters["host"],
+        "port": db_parameters["port"],
+        "account": db_parameters["account"],
+        "schema": db_parameters["schema"],
+        "database": db_parameters["database"],
+        "protocol": db_parameters["protocol"],
+        "timezone": "UTC",
+        "client_session_keep_alive": True,
+    }
+    cnx = snowflake.connector.aio.SnowflakeConnection(**config)
+    try:
+        await cnx.connect()
+        assert cnx.client_session_keep_alive
+    finally:
+        await cnx.close()
+
+
+async def test_keep_alive_heartbeat_frequency(db_parameters):
+    """Tests heartbeat setting.
+
+    Creates a connection with client_session_keep_alive_heartbeat_frequency
+    parameter.
+    """
+    config = {
+        "user": db_parameters["user"],
+        "password": db_parameters["password"],
+        "host": db_parameters["host"],
+        "port": db_parameters["port"],
+        "account": db_parameters["account"],
+        "schema": db_parameters["schema"],
+        "database": db_parameters["database"],
+        "protocol": db_parameters["protocol"],
+        "timezone": "UTC",
+        "client_session_keep_alive": True,
+        "client_session_keep_alive_heartbeat_frequency": 1000,
+    }
+    cnx = snowflake.connector.aio.SnowflakeConnection(**config)
+    try:
+        await cnx.connect()
+        assert cnx.client_session_keep_alive_heartbeat_frequency == 1000
+    finally:
+        await cnx.close()
+
+
+@pytest.mark.skipolddriver
+async def test_keep_alive_heartbeat_frequency_min(db_parameters):
```

Comment:
We should add a test that verify heartbeat is working? Maybe you can check whether a heartbeat http request is sent.

Reply:
updated